sqp/base.py

```diff
@@ -1,172 +1,379 @@
+    """
 
-    hessian_quadratic_p = p @ L_hessian @ p
-    sigma = 1 # hardcoded according to algorithm
+    hessian_quadratic_p = p @ B @ p
+    sigma = 1  # hardcoded according to algorithm
 
     inequality_18_36 = (f_grad @ p + (sigma/2) * hessian_quadratic_p) / ((1-rho) * c_norm)
+
+    # if previous mu fulfills the inequality, use it
     if prev_mu is not None and prev_mu >= inequality_18_36:
         return prev_mu
-    
-    return inequality_18_36 + 1
 
-# should be done (phi_1 function in book)
+    # use a new mu that fulfills the inequality
+    return inequality_18_36 + 0.1
+
+
 def l1_merit(
         f_x: float, 
         mu: float, 
-        c_norm: float):
+        c_norm: float) -> float:
+    """
+    Calculates the merit function at x.
+
+    Args:
+        f_x: Function value at current iterate x
+        mu: Mu that fulfills 18.36
+        c_norm: L1-norm of constraint values at current iterate x
+
+    Returns:
+        Value of merit function
+    """
     return f_x + mu * c_norm
 
-# should be done (D_1(phi_1) function in book)
+
 def l1_merit_directional_gradient(
         f_grad: np.ndarray, 
         p: np.ndarray, 
         mu: float, 
-        c_norm: float):
+        c_norm: float) -> float:
+    """
+    Calculates the directional gradient of the merit function wrt p at x.
+
+    Args:
+        f_grad: Derivative of function at current iterate x
+        p: Direction found with QP
+        mu: Mu that fulfills 18.36
+        c_norm: L1-norm of constraint values at current iterate x
+
+    Returns:
+        Directional gradient of merit function wrt p
+    """
     return f_grad @ p - mu * c_norm
 
 
 def find_lambda_of_qp(
         problem: QuadraticProblem,
         x_solution: np.ndarray) -> np.ndarray:
+    """
+    Computes the lambda vector (Lagrange multipliers) for a given solution x and a quadratic sub-problem
+
+    Args:
+        problem: Quadratic sub-problem
+        x_solution: Solution x of KKT-system for quadratic sub-problem
 
+    Returns:
+        Lambda vector (Lagrange multipliers) for given solution x
+    """
+
+    # start with 0 for all constraints
+    # s.t. all non-active constraints i already have the correct lambda_i=0
     lambda_ = np.zeros(len(problem.constraints))
 
-    g = problem.G @ x_solution + problem.c
+    # find all active constraints at the current point
     active_constraint_mask = np.array([c.as_equality().holds(x_solution) for c in problem.constraints])
     if np.all(~active_constraint_mask):
-        # no active constraints, all lambda=0
+        # no active constraints, early exit with lambda=0
         return lambda_
 
+    # compute derivative of quadratic function
+    g = problem.G @ x_solution + problem.c
+
+    # calculate constraint jacobian for active constraints
+    # note: slicing will work as constraints are always numpy arrays
+    # noinspection PyTypeChecker
     A, _ = combine_linear([c.c for c in problem.constraints[active_constraint_mask]])
```

Comment:
Da kann man auch theoretisch
`A, _ = combine_linear([c.c for c in problem.active_set_at(x_solution)])`
machen. Es is zwar doppelt compute, dafür bräuchte ma ka fancy indexing und ma behält quasi
die Type-requirements.

Sonst wärs vielleicht gut in der `__post_init__` von `MinimizationProblem` die `constraints` zu am numpy array zu formen.